api/src/main/java/marquez/db/migrations/V55_5__BackfillFacets.java

```diff
@@ -0,0 +1,282 @@
+                  WHERE r.uuid = ec.run_uuid
+              )
+              INSERT INTO v55_facet_migration_lock
+              SELECT events_chunk.created_at, events_chunk.run_uuid
+              FROM events_chunk
+              ORDER BY events_chunk.created_at ASC, events_chunk.run_uuid ASC
+              LIMIT 1
+              RETURNING created_at, run_uuid;
+      """;
+
+  @Setter private Integer chunkSize = null;
+
+  @Setter private boolean triggeredByCommand = false;
+
+  @Setter private Jdbi jdbi;
+
+  public int getChunkSize() {
+    return chunkSize != null ? chunkSize : DEFAULT_CHUNK_SIZE;
+  }
+
+  @Override
+  public MigrationVersion getVersion() {
+    return MigrationVersion.fromVersion("55.5");
+  }
+
+  @Override
+  public String getDescription() {
+    return "BackFillFacets";
+  }
+
+  @Override
+  public Integer getChecksum() {
+    return null;
+  }
+
+  @Override
+  public boolean isUndo() {
+    return false;
+  }
+
+  @Override
+  public boolean isBaselineMigration() {
+    return false;
+  }
+
+  @Override
+  public boolean canExecuteInTransaction() {
+    return false;
+  }
+
+  @Override
+  public void migrate(Context context) throws Exception {
+    if (context != null) {
+      jdbi = Jdbi.create(context.getConnection());
+    }
+
+    /**
+     * Workaround to register uuid_generate_v4 function to generate uuids. gen_random_uuid() is
+     * available since Postgres 13
+     */
+    jdbi.withHandle(h -> h.createCall("CREATE EXTENSION IF NOT EXISTS \"uuid-ossp\"").invoke());
+
+    if (getLock(GET_INITIAL_LOCK_SQL).isEmpty()) {
+      // lineage_events table is empty -> no need to run migration
+      // anyway. we need to create lock to mark that no data requires migration
+      jdbi.inTransaction(
+          handle ->
+              handle.execute(
+                  "INSERT INTO v55_facet_migration_lock VALUES (NOW(), uuid_generate_v4())"));
+
+      return;
+    }
+    Optional<MigrationLock> lastExpectedLock = getLock(GET_LAST_LOCK_SQL);
+
+    if (!triggeredByCommand && countLineageEvents() >= BASIC_MIGRATION_LIMIT) {
+      log.warn(
+          """
+              ==================================================
+              ==================================================
+              ==================================================
+              MARQUEZ INSTANCE TOO BIG TO RUN AUTO UPGRADE.
+              YOU NEED TO RUN v55_migrate COMMAND MANUALLY.
+              FOR MORE DETAILS, PLEASE REFER TO:
+              https://github.com/MarquezProject/marquez/blob/main/api/src/main/resources/marquez/db/migration/V55__readme.md
+              ==================================================
+              ==================================================
+              ==================================================
+              """);
+      // We end migration successfully although no data has been migrated to facet tables
```

Comment:
This will leave a lot of database calls that utilize new tables in a bad state. I think better option is to fail the migration - if you won't run the migration, using new version with this successful migration won't work.

Another way, if we really want to suceed this is to remove usage of new tables from db queries, and do them in a future release.

One more question - will running this migration manually even work if we succeed here? I think flyway will see the migration as completed.

Reply:
If you fail flyway, you will not have the empty tables created. A manual run calls a `migrate` method from `V55_5__BackfillFacets.java` class while not starting flyway context, so it is transparent to `flyway`. 

The other issue you raise is splitting the PR into two:
 * first create new tables and give users time to migrate,
 * make use of new tables once the migration is over.

It's add a bit more overhead to a process but may be useful for some users.

Reply:
@mobuchowski your concerns are valid but splitting operation into multiple releases is not a solution as users may run multiple upgrades at once at any point. 

The solution here would be to make code aware of asynchronous migration being finished or not, so that it returns data:
 * from `run_facets`, `job_facets`, `dataset_facets` when migration is over,
 * from `lineage_events` table when migrations is still in progress.

Reply:
That solution sounds good 👍